overlord/snapstate/check_snap.go

```diff
@@ -208,6 +208,13 @@ func checkSnap(st *state.State, snapFilePath string, si *snap.SideInfo, curInfo
 		return err
 	}
 
+	snapName, instanceKey := snap.SplitInstanceName(instanceName)
+	if instanceKey != "" && snapName != s.SnapName() {
+		return fmt.Errorf("cannot install snap %q using instance name %q", s.SnapName(), instanceName)
```

Comment:
why do we only error here if `instanceKey != ""`?

Reply:
Good point.

Reply:
After going a bit back and forth with this one, I've moved the check down, after all callbacks are executed. The assumption is that the callbacks are likely to yield more detailed errors (if any), as for instance happens with `checkGadgetOrKernel()` which returns `cannot replace gadget snap with a different one` while the snap name check would return `cannot install snap "zgadget" using instance name "gadget"`. I'm happy to adjust it further if needed,